javascript/node/events/listening-to-events-just-once.md

```diff
@@ -34,32 +28,39 @@ aspects:
 ---
 ## Content
 
-If more than one uncaught exception was to be caught, the applications equivalent to a `closedown` method, where the application is shutdown, would be triggered more than once, consequently causing issues with the applications shutdown procedure.
+Some applications can include a special behavior that happens whenever it is closed. This behavior can be used to free up resources, close a connection to a databases (or anything else), and for the remainder of this insight we will call it `closedown`.
```

Comment:
One thing I’d mention here is that the focus of the text within the insight is straying too much from its purpose. 

How about this?

```suggestion
An application can end up in various states during its lifetime. Many of these state changes can happen as reactions to events. Although most of the time it's ok to process events as they happen, sometimes we do not want to do this. 

For example, if we had a backend server that accepts payments from a user, we want to make sure that each payment from a user in some exact conditions (for instance when buying a specific product at a specific time) happens only once such that we do not accidentally charge the user multiple times.

Another example could be to include some special behavior for our server that happens only once for any amount of uncaught exceptions. This special behavior could be to shutting down a database connection to free up resources, saving exception logs to a safe location, etc. 

In the examples below, we will demonstrate how to handle this with a callback we'll call `closedown` and a special method of the `EventEmitter` called `once`.
```

Reply:
I guess these scenarios can be considered **non-idempotent** operations. However, this categorization alone doesn't justify or entail the technique we talk about in the insight.

For example, if we talk about handling non-idempotent purchases we most probably won't do that via EventEmitters, right? While there are similitudes in the approach, they might be totally different techniques.

My point is that although all events we will "listen to just once" are non-idempotent, not all non-idempotent events should or must be handled this way.

Hence, I think this defeats that writing guideline that we pseudo-imposed where we try to keep the number of concepts introduced in an insight as low as `1`.

Reply:
You mean **idempotent**, not **non-idempotent**, right? 

As far as implementation of purchases using an EventEmitter, it can certainly happen, implementation is up to the implementor (but also the idea here was more so to think of events generally, not just tied to the api of the EventEmitter).

But I agree 100% that we can move idempotency into footnotes (or even remove it entirely) since it can be scary for a beginner. 

My goal was to suggest another more realistic/familiar idea (most people are familiar with online payments) besides the uncaughtException idea, which felt obscure and edge-casey 🤷‍♂

Reply:
these operations (e.g payments, db closedown etc.) are non-idempotent. they'd be idempotent if they'll run multiple times & produce the exact same result (in terms of state too), right?

about handling rest endpoints as event emitters it def can happen - that's what we do with the GitHub curriculum integration webhooks. but honestly, that only added complexity for us 

I do think too that we could give another example to illustrate the point. lmk what you think of my insight suggestion below - we could append to it another such example

Reply:
So my point was that making `N` (where `N` > 0) payments under equivalent conditions would create only 1 payment. 

As in `F(F(F(F...F(x)...) === F(x)` (one invocations does the same thing as more than one)

That's a safe design in a distributed system. The idea is that, let's say some bank UI let's the user click 50 times on a "Pay my order button", we don't want that order to be paid for 50 times though, right? Ideally the payment should be processed only once no matter the amount of clicks (as long as it's clicked at least once).

Similarly, closing the DB once or N times should always lead to a single closing of the DB.

In both cases, no matter how many times the equivalent event happens, the result is unaffected beyond the first event, hence idempotency.

Reply:
@kapnobatai137 just an FYI, we should go with [@Loopiezlol's suggestion](https://github.com/enkidevs/curriculum/pull/1778#discussion_r366938518), disregard the tangential chatter in this thread lol

Reply:
okay 😆